enhancements/network/ovs-dynamic-cpu-pinning.md

```diff
@@ -0,0 +1,201 @@
+---
+title: ovs-dynamic-cpu-pinning
+authors:
+  - "@zeeke"
+reviewers:
+  - "@trozet"
+  - "@MarSik"
+  - "@ricky-rav"
+  - "@cgoncalves"
+approvers:
+  - "@trozet"
+api-approvers:
+  - "None"
+creation-date: 2023-03-31
+last-updated: 2023-03-31
+tracking-link:
+  - https://issues.redhat.com/browse/CNF-5977
+---
+
+# OVS Dynamic CPU Pinning
+
+## Summary
+
+Allow OVS daemons to run on reserved CPUs if the network workload requires
+cycles and non-reserved CPU are idle.
+
+## Motivation
+
+OVS runs on reserved CPUs when an OCP node is configured with [cpuManagerPolicy](https://docs.openshift.com/container-platform/4.12/scalability_and_performance/using-cpu-manager.html) set to `static`, and the kubelet is configured with [`reservedSystemCpus`](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list).
+
+Typically the users want to minimize the amount of reserved CPUs and leave 
+as many available CPUs as possible for business-related workloads.
+
+We have seen in some customer cases that it is possible to disrupt both the node
+and the cluster when the networking load rises, because OVS is CPU bound by the 
+limited number of system-reserved cores, and networking performance suffers.
+
+### User Stories
+
+> As an Openshift cluster administrator
+> I want to maximize my network dataplane bandwidth to all available cores in a dynamically reserved CPU environment
+> so that the cluster can endure a high network workload.
+
+### Goals
+
+Every cluster node that is configured with static CPU manager policy and
+a low number of `reservedSystemCpus` does not experience any excessive latency 
+increase while dealing with high network loads.
+
+### Non-Goals
+
+The logic used by Node Tuning Operator to decide if and when to enable the feature
+are out of this enhancement's scope.
+
+## Proposal
+
+As of today, a set of CPUs can be reserved for system workload (i.e. other than Kubernetes workload)
+by specifying the `--reserved-cpus` kubelet parameter (or `reservedSystemCpus` via configuration file).
+
+The main idea to resolve the issue described above is to allow OVS 
+to spread out to other cpus outside the reserved cpu pool. However 
+a care must be taken to not interfere with cpus that are assigned 
+to latency sensitive workloads.
+
+### Workflow Description
+
+The feature is triggered by [Node Tuning Operator](https://github.com/openshift/cluster-node-tuning-operator) (NTO) when it detects the 
+performance conditions to turn it on. The logic NTO uses to activate this 
+feature is out of this enhancement's scope. 
+The following steps will be implemented to turn the feature on:
+- NTO's daemonset moves the OVS daemons to a systemd slice different than the default one. 
```

Comment:
Due to changes in 4.13 related to cpu balancing (https://github.com/openshift/kubernetes/commit/461dce542583a54a1a16e3cff3e3fddebd0907b5) NTO will make sure that OVS and related services will stay in a slice that has theoretical (see note 1) access to all cpus and that has cpu balancing disabled (see note 2). It will be either the root cgroup or a nested one like ovs.slice.

Note 1 about cpu access: The slice will allow access to all cpus, but the OVN-K logic will use the sched_setaffinity to restrict OVS components to proper cpus. This is the same method that systemd uses nowadays. The above mentioned OCP patch related to cgroups would have moved OVS into the `system` slice that is restricted to the reserved cpus if NTO have not explicitly overriden this for OVS.

Note 2 about cpu balancing: CFS cpu domain balancing behavior can be a bit tricky to grasp, so an explanation is needed here. OVS will be in a cgroup that has balancing *disabled* via `cpuset.sched_load_balance = 0`. However this merely a request. The kernel CFS scheduler will not disable the balancing for cpus that are being used by any cgroup on the system with balancing enabled. Such as reserved cpus (`system` slice has balancing enabled) and cpus used for burstable and best-effort pods (those have balancing enabled as well). The remaining cpus are being used by guaranteed and cpu pinned containers and OVS will be forbidden to use those via the cpu affinity mechanism described in this proposal.

Reply:
Ok, removing